### PR TITLE
purr 2.0.4 (new formula)

### PR DESCRIPTION
**Pull request by bartekpacia**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

I'm not sure how to add a `test` block – this program requires `adb`, which is located in a  heavyweight Android SDK. And it also requires a connect Android emulator/device.

-----

### CODE REVIEW

File: Formula/p/purr.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def caveats
+    <<~EOS
+      purr requires you to have adb executable in $PATH
+
+      adb is part of Android SDK command line tools. Get them at:
+        https://developer.android.com/studio#command-line-tools-only
+    EOS
+  end
```
Comment (chenrui333) [inline Formula/p/purr.rb:25]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

**Comment by alpiccioni** [comment]
For the test block, if that becomes a problem, I can look into creating a bypass. I think testing purr is going to be very difficult because it exists on top of `fzf`, which is pretty opaque, so testing the UI would be difficult.

What I could do is create a test suite that initializes purr with a mocked ADB client, feeds it some data, and checks if the backing files have initialized correctly. That would at least check if it purr can boot correctly.

File: Formula/p/purr.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    system "make", "purr"
+    bin.install "out/purr"
+  end
+
+  def caveats
+    <<~EOS
+      purr requires you to have adb executable in $PATH
+
+      adb is part of Android SDK command line tools. Get them at:
+        https://developer.android.com/studio#command-line-tools-only
```
Comment (chenrui333) [inline Formula/p/purr.rb:24]: If this tool is for android/adb community, we probably dont need this.

Comment (alebcay) [inline Formula/p/purr.rb:24]: You can reference the `android-platform-tools` cask instead of needing to point at the website.

File: Formula/p/purr.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+class Purr < Formula
+  desc "Versatile zsh CLI tool for viewing and searching through Android logcat output"
+  homepage "https://github.com/google/purr"
+  url "https://github.com/google/purr/archive/refs/tags/2.0.0.tar.gz"
+  version "2.0.0"
```
Comment (chenrui333) [inline Formula/p/purr.rb:5]:
```suggestion
```

File: Formula/p/purr.rb (hunk 2 above)
Comment (alebcay) [reply Formula/p/purr.rb:24]: You can reference the `android-platform-tools` cask instead of needing to point at the website.

https://github.com/Homebrew/homebrew-core/blob/2f7b86a5ba703776a03505881286d8cf91198ad6/Formula/s/scrcpy.rb#L55-L60

**Comment by alebcay** [comment]
For the dependency on `adb` for a meaningful test, one other option is to have a mock/fake `adb` to use with this test. Basically something that generates the same kind of output that `adb` would (in whatever way you are using `adb` here).

One example is how the `scrcpy` formula's `test` block does this:

https://github.com/Homebrew/homebrew-core/blob/2f7b86a5ba703776a03505881286d8cf91198ad6/Formula/s/scrcpy.rb#L64-L105

File: Formula/p/purr.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    <<~EOS
+      purr requires you to have adb executable in $PATH
+
+      adb is part of Android SDK command line tools. Get them at:
+        https://developer.android.com/studio#command-line-tools-only
+    EOS
+  end
+
+  test do
+    assert_match version.to_s, shell_output(bin/"purr -v")
```
Comment (chenrui333) [inline Formula/p/purr.rb:28]: version check is not good enough

Comment (bartekpacia) [reply Formula/p/purr.rb:28]: Yes, I just thought that would be "better than nothing" for now. Fortunately author of `purr` (@alpiccioni) [has an idea](https://github.com/Homebrew/homebrew-core/pull/156118#issuecomment-1836369501) on how to make it work, and [suggestions](https://github.com/Homebrew/homebrew-core/pull/156118#issuecomment-1836374186) from @alebcay here are also very helpful. Thank you!

File: Formula/p/purr.rb (hunk 2 above)
Comment (bartekpacia) [reply Formula/p/purr.rb]: Done - thank you for being so helpful!

Comment (chenrui333) [reply Formula/p/purr.rb]: yeah, cask reference might be a better route. 👍

**Comment by chenrui333** [comment]
looks like linux binary did not get built.

**Comment by alebcay** [comment]
> That's strange – any ideas why it might be?

When I try running this in a Linux docker container, I see:

```
linuxbrew@037ad2ab6831:~$ /home/linuxbrew/.linuxbrew/Cellar/purr/2.0.0/bin/purr
bash: /home/linuxbrew/.linuxbrew/Cellar/purr/2.0.0/bin/purr: /bin/zsh: bad interpreter: No such file or directory
```

My guess is that the ENOENT is from looking for `/bin/zsh`, but this exit code is returned from executing `purr` to the shell, so the test harness _thinks_ its `purr` that's missing, when in fact it's not able to find `zsh` instead. Indeed, it would probably be better to have a shebang of `#!/usr/bin/env zsh` since people might have it installed in different places.

File: Formula/p/purr.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  desc "Versatile zsh CLI tool for viewing and searching through Android logcat output"
+  homepage "https://github.com/google/purr"
+  url "https://github.com/google/purr/archive/refs/tags/2.0.0.tar.gz"
+  sha256 "f4e6c3dcdd386da7d3664fe6c947e9cfdb2e3362ca8c78e47d0445d94d471246"
+  license "Apache-2.0"
+  head "https://github.com/google/purr.git", branch: "main"
+
+  depends_on "make" => :build
+  depends_on "fzf"
+  depends_on "zsh"
```
Comment (alebcay) [inline Formula/p/purr.rb:11]: If there is not a need for some bleeding-edge `zsh` features, I would suggest `uses_from_macos "zsh"` instead of `depends_on "zsh"`.

This allows depending on the macOS-provided copy of `zsh` on macOS, which may not necessarily be the newest (currently 5.8.1 on my Monterey machine, which is the oldest macOSversion that we package for at the moment - possibly newer `zsh` on newer macOS versions). On Linux, the `uses_from_macos` will still cause Homebrew `zsh` to be installed, same behavior as `depends_on "zsh"`.

File: Formula/p/purr.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+class Purr < Formula
+  desc "Versatile zsh CLI tool for viewing and searching through Android logcat output"
+  homepage "https://github.com/google/purr"
+  url "https://github.com/google/purr/archive/refs/tags/2.0.0.tar.gz"
+  sha256 "f4e6c3dcdd386da7d3664fe6c947e9cfdb2e3362ca8c78e47d0445d94d471246"
+  license "Apache-2.0"
+  head "https://github.com/google/purr.git", branch: "main"
+
+  depends_on "make" => :build
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply Formula/p/purr.rb:9]: I tried `uses_from_macos "make"` but it says:

> `uses_from_macos` should only be used for macOS dependencies, not make.

[the remark this answers is not in the document]
Comment (bartekpacia) [reply Formula/p/purr.rb:9]: There's `bsdmake` formula in homebrew though.

[the remark this answers is not in the document]
Comment (alebcay) [reply Formula/p/purr.rb:9]: To be clear, you shouldn't need any dependency line for `make` at all, no need for `depends_on` or `uses_from_macos` (the suggestion to use `uses_from_macos` applies to the `zsh` dep only).

[the remark this answers is not in the document]
Comment (bartekpacia) [reply Formula/p/purr.rb]: Thanks for clarifying 👍🏻

File: Formula/p/purr.rb (hunk 4 above)
Comment (chenrui333) [reply Formula/p/purr.rb:27]: we still need some test improvements on this new formula :)

Comment (bartekpacia) [reply Formula/p/purr.rb:27]: yeah definitely, I'll write a better test tomorrow (if time permits) or on Monday. Thank you for your support!

File: Formula/p/purr.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      You can install adb from Homebrew Cask:
+        brew install --cask android-platform-tools
+    EOS
+  end
+
+  test do
+    puts("bin: #{bin}")
+    assert_match version.to_s, shell_output(bin/"purr -v")
+
+    assert_match version.to_s, shell_output(bin/"purr -A #{bin}/adb_mock")
```
[the remark this answers is not in the document]
Comment (alebcay) [reply Formula/p/purr.rb:35]: Your general approach of building `adb_mock` during the build phase is a common formula pattern (perhaps keep a comment denoting that it's used in the test).

If the intent is to keep it only for testing (not really meant to be run by your end users), you can install it into `pkgshare` (which amounts to `#{share}/purr`). This allows us to keep it for later use but not adding it into a place that would be in users' `PATH`.

```ruby
system "make", "adb_mock", "OUTDIR=#{pkgshare}"
chmod 0755, pkgshare/"adb_mock" # if you still need to make it executable
```

Then using this from the test block looks something like:

```ruby
assert_match version.to_s, shell_output(bin/"purr -A #{pkgshare}/adb_mock")
```

[the remark this answers is not in the document]
Comment (bartekpacia) [reply Formula/p/purr.rb]: Thanks a lot! This worked.

I have a question: why didn't `# bin.install "out/adb_mock"` make `adb_mock` executable?

[the remark this answers is not in the document]
Comment (alebcay) [reply Formula/p/purr.rb]: Good question, this mostly originates from how this is implemented in Ruby. In this case, `bin` is a `Pathname` object and `install` is a method that acts upon it. All `install` does is ensure the parent directory exists and then moves the file there. This behavior is not specific to `bin.install`; the same logic is re-used for `include.install`, `lib.install`, `pkgshare.install`, or some custom path like `(buildpath/"my/sub/dir/path").install`, etc.

There's no logic in it that would explicitly mark the installed files as executable; we rely on the upstream build system (usually autotools, cmake, etc.) to set the executable bit on executable files when appropriate.

[the remark this answers is not in the document]
Comment (alpiccioni) [reply Formula/p/purr.rb]: It seems so. The Ruby docs for Pathname.install (https://rubydoc.brew.sh/Pathname.html#install-instance_method) only refer to moving the file, so either the ruby docs are wrong or the cookbook is worng.

**Comment by bartekpacia** [comment]
I'm having a problem writing this test.

> I would suggest testing by using purr -A $FULL_ADB_STUB_PATH, followed by sending an Esc key after a couple seconds. The return code should indicate whether purr correctly booted and closed.

How should I do that? I tried writing some Ruby code

<details><summary>try</summary>

```ruby
require 'pty'
require 'expect'

# Start the interactive process
cmd = '/opt/homebrew/bin/purr -A /opt/homebrew/share/purr/adb_mock'
r, w, pid = PTY.spawn(cmd)

puts pid

puts "Process status after 0 seconds:"
system("ps", "-o", "pid=", "-p", pid.to_s)

# Wait for 2 seconds
sleep 2

puts "Process status after 2 seconds:"
system("ps", "-o", "pid=", "-p", pid.to_s)

# Send the ESC key to the process
w.puts "\e"
# Close the write end of the PTY
w.close

puts "Process status after sending ESC:"
system("ps", "-o", "pid=", "-p", pid.to_s)
```

</details> 

but it didn't work - i.e. `purr` did not exit. Sorry, I'm not that great of a terminal afficionado.
 
I also found tools like [charmbracelet/gum](https://github.com/charmbracelet/gum) and [`expect(1)`](https://linux.die.net/man/1/expect), but I'm not sure if they're a right fit here.

**Comment by alpiccioni** [comment]
> I'm having a problem writing this test.
> 
> > I would suggest testing by using purr -A $FULL_ADB_STUB_PATH, followed by sending an Esc key after a couple seconds. The return code should indicate whether purr correctly booted and closed.
> 
> How should I do that? I tried writing some Ruby code
> 
> try
> ```ruby
> require 'pty'
> require 'expect'
> 
> # Start the interactive process
> cmd = '/opt/homebrew/bin/purr -A /opt/homebrew/share/purr/adb_mock'
> r, w, pid = PTY.spawn(cmd)
> 
> puts pid
> 
> puts "Process status after 0 seconds:"
> system("ps", "-o", "pid=", "-p", pid.to_s)
> 
> # Wait for 2 seconds
> sleep 2
> 
> puts "Process status after 2 seconds:"
> system("ps", "-o", "pid=", "-p", pid.to_s)
> 
> # Send the ESC key to the process
> w.puts "\e"
> # Close the write end of the PTY
> w.close
> 
> puts "Process status after sending ESC:"
> system("ps", "-o", "pid=", "-p", pid.to_s)
> ```
> 
> but it didn't work - i.e. `purr` did not exit. Sorry, I'm not that great of a terminal afficionado.
> 
> I also found tools like [charmbracelet/gum](https://github.com/charmbracelet/gum) and [`expect(1)`](https://linux.die.net/man/1/expect), but I'm not sure if they're a right fit here.

My Ruby is unfortunately quite weak, so I'm unsure if anything you did here is explicitly wrong. I know `fzf` doesn't like not having a TTY, and it might hook into the Ruby fake TTY in strange ways. It might also not be working at all and hiding some failures.

I think the easiest way forward might be for me to create a test function for `purr` natively that can execute on a standard terminal without outside interference and exits correctly. I think it's a true statement that all the business logic for `purr` happens before `fzf` begins execution, so I think I can just create a command line parameter that prevents execution of `fzf` and bypasses this problem.

I should also create some basic testing to make adb is correctly hooking into purr while I'm at that...

Give me a day or so, I'll get back to you with a solution.

**Comment by alpiccioni** [comment]
https://github.com/google/purr/pull/12 may be a solution. Please check if it works!

File: Formula/p/purr.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (alebcay) [reply Formula/p/purr.rb]: Good catch on the cookbook, I'll take a closer look at that and propose a change if needed.

**Comment by alpiccioni** [comment]
From conversation on purr repo, release 2.0.2 with the file tester should solve the problem.

**Comment by bartekpacia** [comment]
[conversation link](https://github.com/google/purr/pull/12)

Whoa, something strange is going on – this line:

https://github.com/Homebrew/homebrew-core/blob/72d4987c631870a8c67c8500d0ab022b29abbd07/Formula/p/purr.rb#L33

never completes.

I run it like this:

```
$ HOMEBREW_NO_INSTALL_FROM_API=1 brew test purr
==> Testing purr
==> /opt/homebrew/Cellar/purr/2.0.2/bin/purr -v
==> /opt/homebrew/Cellar/purr/2.0.2/share/purr/file_tester -a /opt/homebrew/Cellar/purr/2.0.2/share/purr/adb_mock -p /opt/homebrew/Cellar/purr/2.0.2/bin/purr
```

but when I run the command directly in `zsh`:

```
$ /opt/homebrew/Cellar/purr/2.0.2/share/purr/file_tester -a /opt/homebrew/Cellar/purr/2.0.2/share/purr/adb_mock -p /opt/homebrew/Cellar/purr/2.0.2/bin/purr
```

it completes just fine.

**Comment by alebcay** [comment]
I haven't determined the exact cause of the freeze, but I have found:

- The `file_tester` zsh process is exiting, but the IO pipe (stdout) is left open, so `shell_output` is stuck trying to `gets` on it. Maybe it's not receiving an EOF?
- We can learn more by swapping out a few parts of the command that is running, `shell_output("#{pkgshare}/file_tester -a #{pkgshare}/adb_mock -p #{bin}/purr")`:
  - Replacing `#{pkgshare}/adb_mock` with some other command (e.g. `true`) does not cause the script to terminate (a `grep` error is printed, but hangs similarly to the original)
  - Replacing `#{bin}/purr` with some other command (e.g. `true`) _does_ cause the script to terminate (a `grep` error is printed and then exits with non-zero error code, failing the `brew test`)
- Running the printed test command (e.g. `/usr/local/Cellar/purr/2.0.2/share/purr/file_tester -a /usr/local/Cellar/purr/2.0.2/share/purr/adb_mock -p /usr/local/Cellar/purr/2.0.2/bin/purr`) in `irb` by capturing it with backticks causes the same hanging issue
- Running the printed test command and capturing its output in Python (`import subprocess; subprocess.check_output(['/usr/local/Cellar/purr/2.0.2/share/purr/file_tester', '-a', '/usr/local/Cellar/purr/2.0.2/share/purr/adb_mock', '-p', '/usr/local/Cellar/purr/2.0.2/bin/purr'])`) causes the same hanging issue

My conclusion is that something in `purr` itself is still causing hangs when its output (stdout, specifically) is harvested by other programs.

**Comment by alpiccioni** [comment]
> I haven't determined the exact cause of the freeze, but I have found:
> 
> * The `file_tester` zsh process is exiting, but the IO pipe (stdout) is left open, so `shell_output` is stuck trying to `gets` on it. Maybe it's not receiving an EOF?
> * We can learn more by swapping out a few parts of the command that is running, `shell_output("#{pkgshare}/file_tester -a #{pkgshare}/adb_mock -p #{bin}/purr")`:
>   
>   * Replacing `#{pkgshare}/adb_mock` with some other command (e.g. `true`) does not cause the script to terminate (a `grep` error is printed, but hangs similarly to the original)
>   * Replacing `#{bin}/purr` with some other command (e.g. `true`) _does_ cause the script to terminate (a `grep` error is printed and then exits with non-zero error code, failing the `brew test`)
> * Running the printed test command (e.g. `/usr/local/Cellar/purr/2.0.2/share/purr/file_tester -a /usr/local/Cellar/purr/2.0.2/share/purr/adb_mock -p /usr/local/Cellar/purr/2.0.2/bin/purr`) in `irb` by capturing it with backticks causes the same hanging issue
> * Running the printed test command and capturing its output in Python (`import subprocess; subprocess.check_output(['/usr/local/Cellar/purr/2.0.2/share/purr/file_tester', '-a', '/usr/local/Cellar/purr/2.0.2/share/purr/adb_mock', '-p', '/usr/local/Cellar/purr/2.0.2/bin/purr'])`) causes the same hanging issue
> 
> My conclusion is that something in `purr` itself is still causing hangs when its output (stdout, specifically) is harvested by other programs.

Thanks for the investigation. I'll take a look and see if I can determine what's going wrong.

**Comment by alpiccioni** [comment]
Crud, this is a MacOS specific problem.

On Linux, I see:

```
╰─ python                                                                                                                                   ─╯
Python 3.11.5 (main, Aug 25 2023, 07:43:52) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess;
>>>
>>> subprocess.check_output(['/usr/local/google/home/alpic/purr/out/file_tester', '-a', '/usr/local/google/home/alpic/purr/out/adb_mock', '-p', '/usr/local/google/home/alpic/purr/out/purr'])
b'Serial was loaded.\nInput stream is verbose.\nThread IO working.\nPreview set to instruction.\nModes set to correct starting states.\nVerbose cache is correct.\nThreads are in cleanup.\nThreads did actually cleanup.\nFiles looks valid!\n'
>>> exit()
```

```
 irb                                                                                                                                      ─╯
irb(main):001:0> `/usr/local/google/home/alpic/purr/out/file_tester -p /usr/local/google/home/alpic/purr/out/purr -a /usr/local/google/home/alpic/purr/out/adb_mock`
=> "Serial was loaded.\nInput stream is verbose.\nThread IO working.\nPreview set to instruction.\nModes set to correct starting states.\nVerbose cache is correct.\nThreads are in cleanup.\nThreads did actually cleanup.\nFiles looks valid!\n"
irb(main):002:0> exit
```

On MacOS I can repro. There's something going wrong with the threading. Using some binary search, I can link it to when `__purr_background_handler` launches. What seems to be happening is that the background handler can't launch correctly in the background, and seems to be getting foregrounded, which is blocking (which is correct behavior, if it was in the background).

You can test this out by commenting out the code inside `__purr_background_handler` or the call on line 1685 in the compiled purr and running the file tester in irb or Python. It won't run correctly, but it will exit. Replacing the code with a sleep statement makes it sleep as well.

This is only happening when I'm running `irb` and `python` in MacOS. Shell in MacOS works fine, and everything works in Linux. The background handler is unfortunately a decent amount of the purr business logic, so it's kinda important that it works. 

Do either of you know if there's something specifically in MacOS when running Python or irb that doesn't allow background tasks?

**Comment by alebcay** [comment]
I'm not sure exactly what aspect is breaking, but it probably has something to do with job control when running non-interactively on macOS. Might be related to some differences in how process lifecycle works in macOS vs Linux.

I added the following line to `file_tester` after the `eval` that backgrounds `purr`:

```
background_pid=$!
echo $background_pid 1>&2
```

And I also added the same into `purr`, right after the background handler is started (I put it inside the subshell parentheses).

So running it results in output like (due to stdout being captured):

```
1234
5678
```

And from here, it would hang. After a minute or so, I found that pid 1234 had the `zsh` process (running `file_tester`) that started it as its parent, and it had terminated. But 5678 was still running, and had a parent of `launchd` (pid 1, basically `init` in macOS).

Upon manually killing 5678, everything returned and the test completed successfully. So my guess is that something is going on with how `purr` is exiting but somehow the background process is not getting reaped (possibly related to the fact that the `purr` foreground process is not the background process' parent).

**Comment by alpiccioni** [comment]
Fascinating, this is a regression then. I definitely fixed this issue before. Thanks for the investigation!

I'll go find out what I've done to bring it back...

**Comment by alpiccioni** [comment]
tl;dr: Try purr 2.0.3...

**A yak shaving expedition:**

Here's the issue; I can tell all the tasks are getting ended. If I comment out the `__purr_start_streams` method, the only task should be `__purr_background_handler`. 

If I then run `make clean && make && ./out/purr; ps ax | grep zsh;`, I get:

```
16406 s003  S      0:06.77 -zsh
50856 s003  SN     0:00.00 zsh ./out/purr
50900 s003  S+     0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox zsh
50901 s003  UN     0:00.00 (zsh)
```

So clearly, something is getting left behind. Now, if I comment out the background handler, I get:

```
16406 s003  S      0:06.82 -zsh
51327 s003  S+     0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox zsh
```

Ok, so what is causing that? Process substitution. This line `done < <(/usr/bin/tail -f $purr_background_handler_cache 2>/dev/null)` seems to clean itself up on Linux, but not on Mac. I can actually change the process substitution to just a file description, and it seems to work fine.

Changing this to a named pipe instead fixes the problem. I'm still unsure why the problem exists, but I'm here to solve problem not understand them...

**Sidenote - Leaking Tasks* 

When you run purr with an actual adb client, it seems to clean up just fine (mostly, see second part). The mock adb tasks sleep forever, and don't seem to clean up. That's fine, I can just make them not sleep forever, which more-or-less solves the problem. I'm not super-concerned with why this is happening, since it's not happening with real adb.

**Comment by bartekpacia** [comment]
Wow, what an investigation - huge kudos to both of you.

The test works locally now... except for [mismatch between git tag and `purr -v`](https://github.com/google/purr/blob/f6d0eac8d0bba63b210f94234475b47fc82144cf/src/shell/argument_parsing.sh#L19) :D

**Comment by alpiccioni** [comment]
Pushed it to 2.0.4 with a correct version 😅 

I should really automate that. I'm sure a tool exists.

**Comment by bartekpacia** [comment]
Awesome! We're getting closer and closer.

BTW, just now I realized that @alebcay is the creator of https://github.com/alebcay/awesome-shell – really cool!

File: Formula/p/purr.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      At runtime, adb must be accessible from your PATH.
+
+      You can install adb from Homebrew Cask:
+        brew install --cask android-platform-tools
+    EOS
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/purr -v")
+    shell_output("#{pkgshare}/file_tester -a #{pkgshare}/adb_mock -p #{bin}/purr")
```
Comment (alebcay) [inline Formula/p/purr.rb:33]: The command that `shell_output` is running is returning a nonzero exit code. Is this expected? If so, you can denote the expected exit code as a second argument to `shell_output`, e.g. `shell_output("#{pkgshare}/file_tester -a #{pkgshare}/adb_mock -p #{bin}/purr", 1)`.

Also, `shell_output` only captures the output, it doesn't do anything with it. You can `assert_match` it to some expected stdout output like is done in the preceding line with the version check.

If you don't care about the command line output (if a non-zero exit code would be thrown on error so no need to check/assert on output), you can also use the `system pkgshare/"program", "arg1", "arg2", "arg3"` form to just execute it without output capture. Note that this will fail if the exit code is nonzero though (there's no way to set the expected exit code of a `system` call, it's assumed to be 0).

Comment (alebcay) [reply Formula/p/purr.rb:33]: Also, separately, please make sure the version in the commit message matches the version being shipped in the formula (I know it's easy to overlook).

Comment (bartekpacia) [reply Formula/p/purr.rb:33]: > The command that shell_output is running is returning a nonzero exit code on a few CI machines. Is this expected?

Nope, this is not expected. Looks like a bug

Comment (bartekpacia) [reply Formula/p/purr.rb:33]: > make sure the version in the commit message matches the version being shipped in the formula

fixed, thank you

File: Formula/p/purr.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      At runtime, adb must be accessible from your PATH.
+
+      You can install adb from Homebrew Cask:
+        brew install --cask android-platform-tools
+    EOS
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/purr -v")
+    system pkgshare/"file_tester", "-a #{pkgshare}/adb_mock", "-p #{bin}/purr"
```
Comment (alebcay) [inline Formula/p/purr.rb:33]:
```suggestion
    system pkgshare/"file_tester", "-a", pkgshare/"adb_mock", "-p", bin/"purr"
```

Assuming that this doesn't break anything, from a style perspective we'd prefer to have the args fully separated out.

**Review by alebcay — COMMENTED** [review]
I think this is almost ready - just one minor spot to look at and address if possible.

File: Formula/p/purr.rb (hunk 9 above)
Comment (bartekpacia) [reply Formula/p/purr.rb]: should be okay

**Review by alebcay — APPROVED** [review]

**Comment by bartekpacia** [comment]
great:D

can you merge this @alebcay?

**Review by fxcoudert — APPROVED** [review]
